src/smt/smt_engine.cpp

```diff
@@ -767,7 +771,7 @@ void SmtEngine::defineFunction(Expr func,
 {
   SmtScope smts(this);
   finalOptionsAreSet();
```

Comment:
Could the state handle calling `finalOptionsAreSet()`? I think that it could be argued that the state determines when the final options must be set.

Reply:
Yeah I think this is a good suggestion, I'll make a few more changes.